File: katas/typescript/src/CleanCode/functions/functions-good.ts
// Cumplimiento de Clean Code: Funciones Bien Diseñadas
// ✅ Solución: Una cosa por función, pequeñas, mismo nivel de abstracción, sin efectos secundarios

interface User {
  name: string;
  score: number;
}

class UserScoreProcessor {
  private static readonly FAST_TIME_BONUS = 10;
  private static readonly FAST_TIME_THRESHOLD_SECONDS = 60;

  // ✅ Función hace UNA cosa: orquestar el procesamiento del puntaje
  // ✅ Mismo nivel de abstracción
  public processUserScore(user: User, newScore: number, timeInSeconds: number): void {
    const bonus = this.calculateBonus(timeInSeconds);
    const finalScore = user.score + newScore + bonus;

    this.updateScore(user, finalScore);
    this.sendNotification(user);
    this.logActivity(user, newScore);
  }

  // ✅ Función pequeña que hace UNA cosa: calcular bonus
  private calculateBonus(timeInSeconds: number): number {
    if (timeInSeconds < UserScoreProcessor.FAST_TIME_THRESHOLD_SECONDS) {
      console.log(`Aplica bonus de ${UserScoreProcessor.FAST_TIME_BONUS} puntos por tiempo rápido.`);
      return UserScoreProcessor.FAST_TIME_BONUS;
    }
    return 0;
  }

  // ✅ Función pequeña que hace UNA cosa: actualizar puntaje
  private updateScore(user: User, finalScore: number): void {
    user.score = finalScore;
  }

  // ✅ Función pequeña que hace UNA cosa: enviar notificación
  private sendNotification(user: User): void {
    console.log(`Notificación enviada a ${user.name}: Tu nuevo puntaje es ${user.score}`);
  }

  // ✅ Función pequeña que hace UNA cosa: registrar actividad
  private logActivity(user: User, newScore: number): void {
    console.log(`LOG: ${user.name} obtuvo ${newScore} puntos. Puntaje final: ${user.score}`);
  }
}

// Uso
console.log("=== Cumplimiento de Funciones en Clean Code ===");
const user: User = { name: "Alex", score: 100 };
const processor = new UserScoreProcessor();
processor.processUserScore(user, 50, 30);
console.log("Puntaje final del usuario:", user.score);

export { UserScoreProcessor, User };

File: katas/python/CleanCode/functions/functions_good.py
# Cumplimiento de Clean Code: Funciones Bien Diseñadas
# ✅ Solución: Una cosa por función, pequeñas, mismo nivel de abstracción, sin efectos secundarios

from typing import Dict, TypedDict


class User(TypedDict):
    name: str
    score: int


class UserScoreProcessor:
    FAST_TIME_BONUS = 10
    FAST_TIME_THRESHOLD_SECONDS = 60

    # ✅ Función hace UNA cosa: orquestar el procesamiento del puntaje
    # ✅ Mismo nivel de abstracción
    def process_user_score(self, user: User, new_score: int, time_in_seconds: int) -> None:
        bonus = self._calculate_bonus(time_in_seconds)
        final_score = user["score"] + new_score + bonus

        self._update_score(user, final_score)
        self._send_notification(user)
        self._log_activity(user, new_score)

    # ✅ Función pequeña que hace UNA cosa: calcular bonus
    def _calculate_bonus(self, time_in_seconds: int) -> int:
        if time_in_seconds < self.FAST_TIME_THRESHOLD_SECONDS:
            print(
                f"Aplica bonus de {self.FAST_TIME_BONUS} puntos por tiempo rápido.")
            return self.FAST_TIME_BONUS
        return 0

    # ✅ Función pequeña que hace UNA cosa: actualizar puntaje
    def _update_score(self, user: User, final_score: int) -> None:
        user["score"] = final_score

    # ✅ Función pequeña que hace UNA cosa: enviar notificación
    def _send_notification(self, user: User) -> None:
        print(
            f"Notificación enviada a {user['name']}: Tu nuevo puntaje es {user['score']}")

    # ✅ Función pequeña que hace UNA cosa: registrar actividad
    def _log_activity(self, user: User, new_score: int) -> None:
        print(
            f"LOG: {user['name']} obtuvo {new_score} puntos. Puntaje final: {user['score']}")


if __name__ == "__main__":
    # Uso
    print("=== Cumplimiento de Funciones en Clean Code ===")
    user: User = {"name": "Alex", "score": 100}
    processor = UserScoreProcessor()
    processor.process_user_score(user, 50, 30)
    print(f"Puntaje final del usuario: {user['score']}")

    # ✅ Si necesito cambiar _calculate_bonus() o _send_notification(), cambio una sola vez
    # ✅ Nuevas funcionalidades se agregan fácilmente
    # ✅ Consistencia garantizada
